### ISSUE TITLE
Access on checkbox from outside of the grid

### ISSUE THREAD

**Issue by lsaintmaxent** [Nayjest/Grids#53]
Hey men, me again :)

Is it possible to access to my column of checkboxes with a all check, all un check and print ids buttons of raw checked by the user ?

Really appreciate some help.

Have a nice day,
Louis.

**Comment by Nayjest** [comment]
If you need it on server-side,
1. Check that all checkboxes has unique "name" attributes (you can generate it based on some record id's received from DataRow  via second argument of column callback)
2. Just use Input::get('some_concrete_checkbox_name') somewhere in controller.

But if you incapsulate checkboxes processing logic in grid component, then it's better to use InputProcessor class from Grids package and give names to your inputs based on grid name.

Some example code (not tested):
```php
$column->setCallback(function($val, DataRow $row) use ($grid) {
   $id = $row->getCellValue('some_id_column');
   $name = $grid->getConfig()->getName() . "[checkbox_{$id}]";
   $value = $grid->getInputProcessor()->getValue("checkbox_{$id}", 'SOME_DEFAULT_HERE');
   return "<input type = 'checkbox' name = '$name' ...
    ...and so on
}
```

**Comment by lsaintmaxent** [comment]
Thanks a lot man, I'm trying to get this problem fix, i'll keep you in touch.

Thanks for the support :)

**Comment by lsaintmaxent** [comment]
Thanks a lot man, I'm trying to get this problem fix, i'll keep you in touch.

Thanks for the support :)

**Comment by Nayjest** [comment]
No problem )

**Comment by lsaintmaxent** [comment]
the render of grid checkboxes is ok but i can't get the value in my controller of input::get('checkbox_name').

What i have done is : 
- My grid is rendered with a unique name on each checkbox
- in my view, i have something like : 
<a class="fonction" href="/extranet/public/print_checkboxes" >print checkboxes</a>
{!! $grid !!}

- In my fonction print_checkboxes(Request $request)
i try to access to $request->input('unique_checkbox_name') but the return variable is not set and if i put the grid in the Session::put('grid', $grid) before render the grid, i still can't access to input value.

Is it a post or get problem ? (maybe request is not set if access via a http get method).

Could you help me to understand how should it work ? (i'm pretty new to L5 so i don't relly understand the unseen mechanisms)

Thanks a lot

**Comment by Nayjest** [comment]
You definitely need to clarify your question (ask something more concrete, please :) ), show some code examples to make it possible for me to help you.

Probably this could help:
http://www.html-form-guide.com/php-form/php-form-checkbox.html

Laravel's Input::get() doesn't adds any specific magic that can change that behavior.
Exception is multidimentional input: You can use Input::get('a.b') to extract $_GET['a']['b'] value if I'm not mistaken.

**Comment by lsaintmaxent** [comment]
Hmm yeah it's definitly not clear in my head xD

So let's start from scratch.

I'have render in my grid a column of checkboxes to select many raw to print in a pdf.
Here's the code of the column : 
(new FieldConfig)
	->setName('mod_id')
	->setLabel('imprimer etiquette')
	->setCallback(function ($val){
	return "<input type=\"checkbox\" name=\"impression[]\" value=\"".$val."\" id=\"".$val."\" />";
})
,
(i also tried with a unique name, as name=\"checkbox_".$val."\" ... but that didn't work).

Then i render the grid in my view with a link ta a function that should (in the future) print all stickers of raw with checked checkbox:
 <a class="fonction" href="/extranet/public/imprimer_etiquettes" >Imprimer les étiquettes</a>
<center>{!! $grid !!}</center>

And in this function (imprimer_etiquette), i'm trying to access to these checkboxes cause they have in value field the id of mods which need to get stickers printed.
	public function imprimer_etiquettes(Request $request){
		$checkboxes = 	$request->input('impression[]');
		if(isset($checkboxes)) echo "youpi";
		else echo "ohh :\'(";
	}

But i can't access to impression[]. i also tryed with impression and checkbox_[an id of my DB here] and i can't get it working.

I also tried to store $grid in the session before render and get it back in this function with Session::get('grid') but the return is always empty :/

I'm pretty lost right now xD Thanks a lot for your support !

**Comment by Nayjest** [comment]
>> Here's the code of the column : 
...
->setCallback(function ($val){
return "";
})

I still don't  see how checkboxes are rendered )

**Comment by lsaintmaxent** [comment]
oopsy :)
return "<input type=\"checkbox\" name=\"impression[]\" value=\"".$val."\" id=\"".$val."\" />";

**Comment by Nayjest** [comment]
What will print 
```php 
dd(\Input::get('impression')); 
```
on form submit?

* Add this line to beginning of your controller action  **after opening page  with grid in browser** and then submit form. To submit form, you need to have a submit button inside grid.

**Comment by Nayjest** [comment]
I hope, you don't added there another form? 
Grid already contain one form and there sould be no other forms.

**Comment by lsaintmaxent** [comment]
my first idea was to added another form like : 
{{Form::open(...)}}
{{$grid}}
{{Form::close()}}
but this didn't worked and new issues appear with filters xD

By the way, when the button is inside the grid, i can get my input correctly : 
Array ( [0] => 1 [1] => 3 [2] => 4 ) for the the checkboxes checked :)

So now i just need to put my button in the grid and create a button to check all / uncheck all.

Can i find an exemple somewhere of these button ?

**Comment by Nayjest** [comment]
```php
(new HtmlTag)
    ->setTagName('button')
    ->setAttributes([
        'type' => 'submit',
        'class' => 'btn btn-primary'
])->setContent('Submit form')
```

**Comment by lsaintmaxent** [comment]
Thanks a lot :)

**Comment by lsaintmaxent** [comment]
Just a last question, all is working great thanks to you but i have a last problem about grid form.

I added my function call at the begining of the index() function in the controller.

But my button is generating a pdf, and need an exit() call at the end of the generation so the pdf replace my grid instead of opening a new tab and display itself in this new tab.

Should i fix this using javascript ? or maybe should i call this function after the render ?

Regards,
Louis

**Comment by Nayjest** [comment]
>> But my button is generating a pdf, and need an exit() call at the end of the generation so it's squeezing the return to the view.
Some kind of data export? 
If you generate PDF and return it to browser, it's ok to use exit() considering that Laravel has no suitable functions for ending app.
See how CSV & Excel export components are done. There is also exit(). I'm planning to add something like callable 'finishApp' property to export components that will contain "exit" by default. This will allow to finish application correctly when exit isn't suitable for any reason.

**Comment by lsaintmaxent** [comment]
Yeah, kind of data export (i used FPDF library to generate a pdf of raw data).

The problem is that if the button is in the grid, and submit the grid form, the view will be refresh.
But in the case of my pdf, i wanted to keep grid in actual tab and open a new one with pdf generated.

But if i want to keep grid i can't redirect to another view, and with the exit function, the grid is not rendered (there is only my pdf in the tab).

I would like to use something like html balise " <a href with and target="_blank">" but this will not work because a "<a>" link will not submit the grid form.

So i think i can't solve this problem right now without using some tricky javascript somewhere xD

**Comment by Nayjest** [comment]
Form can have target = "_blank" too.
But  if you use form for filtering, you don't need this attribute.
So, you may have "Filter" or "Refresh" button that is just a submit button for applying filters and "Export" button, that will add hidden input "export_to_pdf" (on server side you will check that input and render PDF instead of grid if input exists). And you need to add onclick handler to "Export button" that will add "target='_blank'" attribute to form and then submit it via JS. Profit!

**Comment by lsaintmaxent** [comment]
Oh wow :/ thanks anyway :)

**Comment by majacirkova1989** [comment]
Hey lsaintmaxent , can you explain to me how you make render of checkbox cause i have the same problem i want on multiple select checkboxes to render (filter) on submit button ?

Thanks Maja

**Comment by lsaintmaxent** [comment]
Yeah, sure,
In your gird configuration create a new column with this :
(new FieldConfig)
	->setName('a_random_unique_id')
	->setLabel('column_name')
	->setCallback(function ($val){
		return "<input type=\"checkbox\" name=\"impression[]\" value=\"".$val."\" id=\"".$val."\" />";
	})
,

Tell me if you have some troubles :)
